### ISSUE TITLE
convautoencoder with traits

### ISSUE THREAD

**Issue by tnn1t1s** [jmsdnns/pallets#6]
In this model:

The image_encoder encodes the input images into a flattened representation.
The trait_encoder encodes the trait information, producing a tensor of the same size as the flattened image representation.
The encoded image and trait data are combined using element-wise addition.
The decoder reconstructs the image from this combined representation.

In cpunks - see https://github.com/tnn1t1s/cpunks-10k/blob/main/cpunks/tutorial/Simple%20Neural%20Net%20Classifier.ipynb

you can get labels here using the .pkl file into a dataframe or the json. we should probably find a better way to do this.

**Comment by jmsdnns** [comment]
We added the labels from cpunks to the VAE in this [notebook](https://github.com/tnn1t1s/pallets/blob/main/nb/AddLabelsToVAE.ipynb)

The first approach was done naively, and we found the output was not great. After reflecting on why, we determined the image data, represented with meaningful one-hot vectors was getting distorted when processed along side the labels data, which is not one-hot encoded. This can be seen towards the bottom of the notebook.

We followed up by introducing a gumbel softmax calculation at the end of each forward pass. This allows the output to be influenced by backpropagation by nature of turning it into a distribution (_which is differentiable_) of values contained in a (k-1) simplex, on which softmax is then run.

A better explanation of how it works will be present in the paper covering the work soon
